### PR TITLE
Adding --ddl-wrapper-comment for diff output

### PR DESCRIPTION
**Pull request by shlomi-noach**
A new flag is introduced, `--ddl-wrapper-comment` (bool, default `false` for backward compatibility), which controls output of `skeema diff` (or `skeema push --dry-run`) command.

When this flag is given, each DDL is wrapped by well formed comments, in the likeness of:
```
-- skeema:ddl:begin
CREATE TABLE `table_0` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(128) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
-- skeema:ddl:end
```

The purpose of this flag is to assist with parsing of `skeema diff` output. As a concrete example, we want to be able to run `skeema diff` and review the statements generated by the tool, on a per-statement basis. When `diff` contains multiple DDL statements (multiple `CREATE`, `ALTER`, `DROP`) it is a difficult task to parse it. Toekizing by semicolon (`;`) is unsafe, because a semicolon can appear as, say, a `VARCHAR`s default value, or in a column comment.

With the above wrapper comments the parsing operation is safe and simple. The wrapper comments text is extremely unlikely to appear as a valid text in a DDL.

---

I hope this PR finds you well. It confirms to existing coding standards. As I looked into a unit test, I found it difficult to follow the docker usage in generating & comparing diffs:

https://github.com/skeema/skeema/blob/94f380a5c46124ee743a221f9b2bb0b67b1c7859/skeema_cmd_test.go#L442-L453

I also noted the above uses `--brief` and so does not compare actual DDL statements; and so I'm unsure how to test wrapper statements. I'm happy to work on this with guidance, as needed.

cc @github/database-infrastructure

### CODE REVIEW

**Comment by evanelias** [comment]
Hi Shlomi, thank you for the PR! The code looks good, but the option naming could be confusing vs the existing [ddl-wrapper](https://github.com/skeema/skeema/blob/master/doc/options.md#ddl-wrapper) option, which configures shelling out to an external program per DDL.

Actually, speaking of [ddl-wrapper](https://github.com/skeema/skeema/blob/master/doc/options.md#ddl-wrapper), I wonder if it might already solve your use-case more cleanly? When running `skeema push`, if you've configured e.g. `ddl-wrapper=/path/to/some/script --ddl {DDL} --schema {SCHEMA}` (or more options as needed) then your external script would be invoked once per DDL statement, without you needing to tokenize them first.

That requires using `skeema push` though, to actually execute the wrapper script, whereas `skeema diff` will just print the generated wrapper command-line without actually running it. If your script does not actually *execute* any of the DDL though, this usage is a bit unusual, so maybe there's a better way to do this. A few questions about your use-case if you're able to share:

* Is this part of an automated pipeline (tokenized DDL needs to be fed to another script)? Or is a human manually reviewing the generated DDL?
* Is the motivation for this requirement just a temporary process (initially evaluating Skeema / confirming its correctness), or a permanent part of a workflow?
* Who/what eventually executes the generated DDL?

re: your test question, I plan to improve the docs there soon. The test suite is powerful and flexible, and makes *running* complicated tests easy, but at the expense of the code being a bit hard to follow. The snippet you highlighted is one of the few places where the STDOUT output is examined and parsed. There's no good helper method for that yet, since most of the integration tests instead evaluate the resulting state of the database and/or filesystem, rather than the STDOUT output.

**Comment by shlomi-noach** [comment]
Hi Evan, thank you for the review! I can elaborate and explain a bit more about our intentions, though much of it is work in progress. First, answering:

> If your script does not actually execute any of the DDL though, this usage is a bit unusual

Naturally this is the situation 😂 . We only want the `diff` without the `push`.

> Is this part of an automated pipeline (tokenized DDL needs to be fed to another script)? Or is a human manually reviewing the generated DDL?

Yes to both.

> Is the motivation for this requirement just a temporary process (initially evaluating Skeema / confirming its correctness), or a permanent part of a workflow?

In the current design, I expect it to be permanent.

> Who/what eventually executes the generated DDL?

Explained shortly, as well as the reasoning for not having `skeema` do this.

The problem we're trying to solve, of course, is schema migration flow. For us, the flow includes:

- the developers designing a change
  - for us this unsurprisingly means a branch/PR in a git/github repo.
- generation of migration statements (`skeema diff`)
- sanity tests: dev env matches prod env, is up to date 
- reviews by various teams
  - possibly reject one or more of the statements.
  - possibly choose a different execution path for a specific change (this change we want at higher priority; that change we want to roll out on replica; this change we want on a different cluster or on two clusters because we're in the middle of some splitting operation)
- analysis of where the change should be applied to. This is more than host/schema: a cluster could be sharded, in which case there's multiple masters to apply the change to.
- running the migration, in a way that is coupled with our infrastructure. Any way I look at this, we need the following:
  - visibility in chat/issue comments on status
  - chatops to view & control the current execution (suspend, cancel)
  - if cancelled, a way to mark the migration as "blocked" until a human resolves it (but this should not affect other migrations that happen to exist in the same diff.
- PR comments, mentions, reviews, chat messages throughout the operation to indicate status.
- To top that, we'd preferably run the control over kubernetes, which means the pod can go down at any moment. As you know, we use `gh-ost` for schema migrations, and `gh-ost` itself does not run on kubernetes; but the service that would kick it is expected to. Our k8s services integrate well with our infrastructure (logging, chatops, monitoring, errors).

In additions we'd need to somehow adapt to an existing system based on ruby/rake, which does not use `.skeema` files or any of its standards. So there's some patchwork here.

The way I look at it, and as an open source developer, if we were to use `skeema` to run `gh-ost`, we'd need to make quite a few changes that are specific to us, and which will not make sense to the general user. Therefore, I prefer running the migrations ourselves, getting `skeema` to generate the `diff` and to make sense of the migration.

Running from k8s means the service that runs the migration needs to be able to go down, and for a new service to pick up from where it left. I don't see that any of this makes sense for `skeema`.

This PR is possibly only change we'd expect to get from `skeema` to be able to operate the work from the outside.
If you feel uncomfortable with the PR that's understandable and we can always work off a fork.

**Comment by evanelias** [comment]
> We only want the `diff` without the `push`.

In this case, I think [ddl-wrapper](https://github.com/skeema/skeema/blob/master/doc/options.md#ddl-wrapper) can already fully solve your problem, without the need for this PR. For example, you could append the DDL statements (with extra wrapping comments) to a log file, by using this configuration in a top-level .skeema file:
```
ddl-wrapper=echo "-- skeema:ddl:begin\n"{DDL}";\n-- skeema:ddl:end" >>{SCHEMA}-ddl.log
```
You'd just have to change your pipeline use `skeema push` instead of `skeema diff`. Don't worry though, it won't actually push anything itself. Due to the configured ddl-wrapper, `skeema push` will just run the configured wrapper script to do the "push", instead of Skeema executing the DDL directly. And in this case the wrapper script just appends the DDL to a log file (foo-ddl.log if the schema name is "foo") without actually running the DDL. The next process in your pipeline can then tokenize that file and handle accordingly.

Alternatively, you could set the ddl-wrapper to a custom application which actually handles the execution, including making automated bot comments in chat/PRs/etc. Ultimately, ddl-wrapper can do whatever you want; Skeema is designed to be able to pass in whatever information is relevant on the ddl-wrapper's command-line. Overall it's up to you whether you want Skeema to execute DDL, or your ddl-wrapper to execute DDL, or some subsequent process in your CI/CD pipeline to execute DDL.

A few thoughts on some of the other bullet points:

> * possibly reject one or more of the statements

The intended workflow for Skeema is that this should happen earlier, at the PR code review stage, rather than during the "deployment" (push) stage. To remove a DDL statement from a PR, the PR creator should adjust their PR (adding a new commit / reverting part of a prior commit) accordingly. 

By design, it's the literally the exact same workflow as if you're rejecting part of an application code PR :) Skeema is an infrastructure-as-code solution, allowing you to use similar idioms around code review, deployment, and release engineering as you would for application code changes. So just like some particularly gnarly code changes need special care, the same may be true of certain high-risk schema changes.

> we'd need to somehow adapt to an existing system based on ruby/rake, which does not use .skeema files or any of its standards.

For development, when working with Rails (for example), the intended workflow is to make changes to dev using standard Rails migrations. A developer can then use `skeema pull dev` to update the schema repo to represent whatever changes were just made in dev, and then commit to a branch, make a PR, etc.

For pushing to prod, any Rails-specific activities (such as updating the rails metadata `schema_migrations` table) could be handled as part of the ddl-wrapper, or handled by the prior or subsequent steps of the CI/CD pipeline.

> * analysis of where the change should be applied to. This is more than host/schema: a cluster could be sharded, in which case there's multiple masters to apply the change to

Skeema has native support for sharding, supporting any arbitrary sharding layout via its [host-wrapper](https://github.com/skeema/skeema/blob/master/doc/options.md#host-wrapper) and [schema](https://github.com/skeema/skeema/blob/master/doc/options.md#schema) options. In combination you can use these to map a single directory to multiple schemas on one mysqld, one schema on multiple mysqld, or multiple schemas on multiple mysqld.

Skeema diffs each shard individually, which means it's totally ok if a few shards are "behind", for example if a failover interrupted a prior schema change operation on one or more shards. This means that once you configure sharding, you can just run `skeema diff` to instantly see if any shards are in the wrong state.

> The way I look at it, and as an open source developer, if we were to use skeema to run gh-ost, we'd need to make quite a few changes that are specific to us, and which will not make sense to the general user. Therefore, I prefer running the migrations ourselves, getting skeema to generate the diff and to make sense of the migration.

That's basically the intention of ddl-wrapper already :) That's the reason I'm hesitant about this PR -- Skeema is already designed to handle arbitrary company-specific customizations by providing flexible hook points.

That said, there's always room for more hooks, and room for more command-line variables for each hook. So I'm definitely open to suggestions, if ddl-wrapper isn't able to accommodate some specific part of your workflow?

**Comment by shlomi-noach** [comment]
Thanks @evanelias for the elaborate answer! I'll look into all of that, starting with `ddl-wrapper` but also into sharding capabilities. Hats off!

**Comment by shlomi-noach** [comment]
(if the `ddl-wrapper` solution works, I'll just go ahead and close this PR)

**Comment by shlomi-noach** [comment]
@evanelias OK cool, I'm good with `--ddl-wrapper`. I ended up with:

`skeema push ci --dry-run --ddl-wrapper='-- skeema:ddl:begin --{DDL}-- skeema:ddl:end --' |  sed -e "s/[\][!] -- skeema:ddl:begin --'/-- skeema:ddl:begin --\n/g" | sed -e "s/'-- skeema:ddl:end --/\n-- skeema:ddl:end --/g"`

Not the prettiest but it will do!

Please feel free to close this PR.

**Comment by shlomi-noach** [comment]
Hmmm actually there's still some quoting I find difficult to get rid of. Even if I run the simplest `--ddl-wrapper={DDL}`, I get:
```
$ skeema diff ci --ddl-wrapper={DDL}
-- instance: 127.0.0.1:3306
\! 'CREATE DATABASE `skeefree_test` CHARACTER SET latin1 COLLATE latin1_swedish_ci'
USE `skeefree_test`;
\! 'CREATE TABLE `sample_data` (
  `id` int(10) unsigned NOT NULL DEFAULT '"'"'0'"'"',
  `hostname` varchar(128) NOT NULL,
  `pid` int(10) unsigned DEFAULT NULL,
  `time_created` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `time_updated` datetime DEFAULT NULL,
  `i` int(11) NOT NULL DEFAULT '"'"'18'"'"',
  `random_hash` char(32) CHARACTER SET ascii DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `random_hash_key` (`random_hash`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8'
\! 'CREATE TABLE `some_table` (
  `id` int(10) unsigned NOT NULL DEFAULT '"'"'0'"'"',
  `hostname` varchar(128) NOT NULL,
  `pid` int(10) unsigned DEFAULT NULL,
  `time_created` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `time_updated` datetime DEFAULT NULL,
  `n` int(11) NOT NULL DEFAULT '"'"'18'"'"',
  `random_hash` char(32) CHARACTER SET ascii DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `random_hash_key` (`random_hash`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8'
2019-09-04 11:17:22 [INFO]  127.0.0.1:3306 skeefree_test: diff complete
```

where 
```
`id` int(10) unsigned NOT NULL DEFAULT '"'"'0'"'"',
```
is something I'm struggling with to un-quote via shell.

**Comment by evanelias** [comment]
It's actually pre-quoted so that it can appear on a shell command-line properly... if you switch to `skeema push` and use a real command-line (even just `echo`), you should find it executes properly :)

```
$ skeema push --ddl-wrapper="echo {DDL}"

-- instance: 127.0.0.1:3306
USE `product`;
\! echo 'CREATE TABLE `some_table` (
  `id` int(10) unsigned NOT NULL DEFAULT '"'"'0'"'"',
  `hostname` varchar(128) NOT NULL,
  `pid` int(10) unsigned DEFAULT NULL,
  `time_created` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `time_updated` datetime DEFAULT NULL,
  `n` int(11) NOT NULL DEFAULT '"'"'18'"'"',
  `random_hash` char(32) CHARACTER SET ascii DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `random_hash_key` (`random_hash`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8'
CREATE TABLE `some_table` (
  `id` int(10) unsigned NOT NULL DEFAULT '0',
  `hostname` varchar(128) NOT NULL,
  `pid` int(10) unsigned DEFAULT NULL,
  `time_created` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `time_updated` datetime DEFAULT NULL,
  `n` int(11) NOT NULL DEFAULT '18',
  `random_hash` char(32) CHARACTER SET ascii DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `random_hash_key` (`random_hash`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

The first STDOUT output is just to tell you what it's running (which is indeed a bit mangled due to shell escaping), but the subsequent STDOUT is from the actual echo command and is correct. If you use `skeema diff` you only get the top one, because `skeema diff` doesn't actually run the DDL wrapper, only `skeema push` does.

To de-dupe the output, in a "real" wrapper you'd process the {DDL} or append it to a file, rather than just echo'ing to STDOUT.

**Comment by evanelias** [comment]
btw overall my recommendation would be to try putting this in a top-level .skeema file:

```
ddl-wrapper=echo "-- skeema:ddl:begin\n"{DDL}";\n-- skeema:ddl:end" >>{SCHEMA}-ddl.log
```

and then use `skeema push` instead of `skeema diff`. I tested that yesterday and it seemed to give the output you're looking for, as long as the next script in the pipeline can then process the resulting log file. (you might need to iterate from there on the ddl-wrapper re: the log file's location, or use other variables supported by ddl-wrapper, etc...)

**Comment by shlomi-noach** [comment]
Ahhh, gotcha! This is an actual `push` that prevents MySQL from actually applying the DDL but rather echo it -- cool!

**Comment by evanelias** [comment]
Exactly :) Does that solve your use-case cleanly then when switching to `push`?

On a side-note, I just realized I need to improve the docs for `ddl-wrapper` and `alter-wrapper`... The docs are admittedly unclear about this point, i.e. if a wrapper is configured, `skeema push` *only* runs the wrapper script and does not otherwise execute the DDL. And in retrospect I probably should have chosen a clearer word than "wrapper" in those option names, but too late to change that aspect of it :)

**Comment by shlomi-noach** [comment]
Cool. This is what I ended up with:
```
skeema push foo --ddl-wrapper='echo "\n-- skeema:ddl:begin\n"{DDL}";\n-- skeema:ddl:end"' |  sed -n '/^-- skeema:ddl:begin/,/^-- skeema:ddl:end/p'
```

An example output for that (after `sed`) is:
```
-- skeema:ddl:begin
ALTER TABLE `sample_data` ADD COLUMN `i` int(10) unsigned DEFAULT '0' AFTER `time_updated`, ADD KEY `i_idx` (`i`);
-- skeema:ddl:end
-- skeema:ddl:begin
ALTER TABLE `some_table` ADD COLUMN `v` varchar(20) NOT NULL DEFAULT '' AFTER `time_updated`;
-- skeema:ddl:end
```

So I'm happy and able to get what I came for 😆 
Please feel free to close this PR.

**Comment by shlomi-noach** [comment]
(or, actually, I can close it...)
